File: problem-solving/_lc-409-easy-longest-palindrome/2-single-pass.go
/*
https://leetcode.com/problems/longest-palindrome/

Complexity
----------
Time: O(n)
Space: O(n)
*/

type void struct{}

func longestPalindrome(s string) int {
    var nullVal void
    set := make(map[rune]void)
    for _, c := range s {
        if _, ok := set[c]; ok {
            delete(set, c)
        } else {
            set[c] = nullVal
        }
    }
    if len(set) > 0 {
        return len(s) - len(set) + 1
    }
    return len(s)
}

File: problem-solving/lc-409-easy-longest-palindrome/2-single-pass.py
"""
https://leetcode.com/problems/longest-palindrome/

Complexity
----------
Time: O(n)
Space: O(n)
"""

class Solution:
    def longestPalindrome(self, s: str) -> int:
        seen = set()
        for c in s:
            if c in seen:
                seen.remove(c)
            else:
                seen.add(c)
        if len(seen) > 0:
            return len(s) - len(seen) + 1
        return len(s)
